Moving leptonSF code to compiled class

amagnan opened the pull request:
- the previous version, with a helper macro, was working fine interactively, but not in crab. Makes more sense to have properly compiled CppWorker. 
- Added possibility to produce loose or tight working points in the leptonSFProducer.py  -> POG still needs to extend to other WP, and compute errors properly -> need source-by-source variations, how it is done now is incorrect I believe (but not saved in output branches at the moment anyway).

amagnan commented:
Related to issue https://github.com/cms-nanoAOD/nanoAOD-tools/issues/108

arizzi commented:
can you fix the conflicts so we may consider for merge? (apologies for the late feedback)

amagnan commented:
I fixed the merge error (removing the python helper code which was moved in proper compiled cpp class) and also I fixed the SF error to be returned by "type", up to the user then to combine properly the different errors in quadrature or keep as separate ....

arizzi commented:
@sroychow can you validate / cross check with the similar code you have?

sroychow commented:
@arizzi My code was updated only for Muons. I will check if the same will work for electrons and get back to you.

sroychow commented:
@arizzi  
Please have a look at 
https://github.com/sroychow/nanoAOD-tools/blob/a1cefbe44f496e4c95ce6e8157383f13c61a9311/python/postprocessing/modules/common/lepSFProducer_v2.py

The module takes as input histograms(depending on working points) to read SF and uncertainties from the MuonPOG trees. It only uses the WeightCalculater helper class. I have also checked the root files  provided for electrons and the module can be used for egamma case as well.

arizzi commented on python/postprocessing/modules/common/lepSFProducer.py:
```diff
@@ -26,18 +55,20 @@ def __init__(self, muonSelectionTag, electronSelectionTag):
-        if "/LeptonEfficiencyCorrector_cc.so" not in ROOT.gSystem.GetLibraries():
-            print "Load C++ Worker"
-            ROOT.gROOT.ProcessLine(".L %s/src/PhysicsTools/NanoAODTools/python/postprocessing/helpers/LeptonEfficiencyCorrector.cc+" % os.environ['CMSSW_BASE'])
+        for library in [ "libCondFormatsJetMETObjects", "libPhysicsToolsNanoAODTools" ]:
```
why are we checking for METObjects?
